### ISSUE TITLE
'Failed to create process' after moving the WinPython directory

### ISSUE THREAD

**Issue by AlexanderYAPPO** [winpython/winpython#329]
I installed external packages using pip from **WinPython Command Prompt.exe** on Win7 64. (**pyramid** in particular which provides **pserve.exe** executable). Suppose the path of WinPython directory is **C:\work_space\MyProgram**. In this case everything works fine even after moving and renaming WinPython directory (except for paths with white spaces of course). The problem appears when I move this WinPython distribution to another PC with Win10 64. If the absolute path stays the same it works, but if path is changed (for instance, **C:\work_space\dir\MyProgram**) I can't run any executable from WinPython Command Prompt. The weird thing is that when a try to run **pserve** (or **pserve.exe**) it simply says: **failed to create process**. Python interpreter still works in all cases, the problem only happens with running exe files. 

Also, I checked all environment variables that are set in env.bat and they seem to be right.

It seems like a bug, but I'm not 100% sure.

**Comment by AlexanderYAPPO** [comment]
I've found out that all *.py files from **python-3.4.4.amd64/Scripts/** directory specify a path to the interpreter and of course it is hardcoded to the previous path.  I was able to run pserve on the local PC because I had another python interpreter in that directory.  If I specify correct path to the interpreter in these *.py files everything should be okay.

But is it how it is supposed to be?

**Comment by stonebig** [comment]
on recent Winpython, you should have a `scripts\make_winpython_movable.bat` that should bring you back to a movable state:
````
@echo off
call %~dp0env.bat
echo patch pip and current launchers for move

%WINPYDIR%\python.exe -c "from winpython import wppm;dist=wppm.Distribution(r'%WINPYDIR%');dist.patch_standard_packages('pip', to_movable=True)"
pause
     
````


Problem is that there is a patch to apply each time you upgrade pip, to keep new installed package in 'relative' path.

So you have also now a `script\upgrade_pip.bat`
````
@echo off
call %~dp0env.bat
echo this will upgrade pip with latest version, then patch it for WinPython portability ok ?
pause
%WINPYDIR%\python.exe -m pip install --upgrade --force-reinstall  pip
%WINPYDIR%\python.exe -c "from winpython import wppm;dist=wppm.Distribution(r'%WINPYDIR%');dist.patch_standard_packages('pip', to_movable=True)
pause
````

**Comment by AlexanderYAPPO** [comment]
How should I do it exactly? That's how I tried to do it. 

I have an installed WinPythonZero. I install `mingwpy` and `alembic` using pip (alembic installation needs mingwpy installed, otherwise it fails). So, all I do is execute
```
pip install -i https://pypi.anaconda.org/carlkl/simple mingwpy
pip install alembic
```
from **WinPython Command Prompt.exe**. I tried to run `script\upgrade_pip.bat` (and `scripts\make_winpython_movable.bat` too) before and after these commands but in both cases renaming WinPython directory led to `failed to create process` error when I tried to execute alembic.

I tried to understand what `scripts\make_winpython_movable.bat` does and It looked like it checks all `*.exe` files from `Script` directory and replaces `#!python.exe` substring there. In my case `initial_content == final_content` so nothing really changes. 

P.S. It could be I use inappropriate mingwpy but I don't think it's an issue.

**Comment by stonebig** [comment]
ok. I thought you had updated pip. which version of WinPython do you use ?

**Comment by AlexanderYAPPO** [comment]
WinPython-64bit-3.4.4.2Zero

**Comment by stonebig** [comment]
the last one,... So I don't know what I can do, you may have to inquire on your own where/why it fails. then, we'll see if Winpython can do something better

**Comment by stonebig** [comment]
you can also experiment the "make_winpython_fix.bat" batch, as a last resort option before starting deep diving on your issue.

**Comment by AlexanderYAPPO** [comment]
Ok, I'll take a look. But why can't we write a script that rewrites paths to the interpreter in all `*.py` files in the `Script` directory? I tried it in my situation and It works but I'm not sure it does in some special cases. If there is no problems then we can add such script.

**Comment by stonebig** [comment]
`it works` == `it solves your issue` ?

**Comment by AlexanderYAPPO** [comment]
Yes, it works so far. Here is the scrips if someone needs it:
```
def change_interpreter_path(python_scripts_path):
    all_files = [python_scripts_path + i for i in os.listdir(python_scripts_path)]
    py_files = list(filter((lambda x: x[-3:] == '.py'), all_files))

    for line in fileinput.input(py_files, inplace=True):
        if re.match('^#\!.*python\.exe$', line) is not None:
            print('#!.\python.exe')
        else:
            print(line, end='')

if __name__ == '__main__':
    change_interpreter_path('WinPython-64bit-3.4.4.2Zero/python-3.4.4.amd64/Scripts/')
```
I can create a pull request on this if you want.

**Comment by stonebig** [comment]
A pull request is welcomed.

**Comment by stonebig** [comment]
by the way , on ".exe" files, the "!.\python.exe" trick didn't work for some packages with subprocess. so I'm not 100% sure of the "all case"... there may be a trace of myu failed experiment in the commit history

**Comment by stonebig** [comment]
as far as I remember, sqlite_bro.exe could survive this change, but another one, maybe wppm.exe couldn't

**Comment by AlexanderYAPPO** [comment]
I searched the pip source code and I found out several interesting things. When pip installs a package and creates an executable file it actually uses `setuptools.command.easy_install.WindowsScriptWriter` for creating an executable.

As I thought, all this executable file does is just fetch an interpreter specified in the `#! ` string and then execute the script in it. 

Easy install documentation describes the process in more details: https://github.com/pypa/setuptools/blob/04d10ff025e1cbef7ec93a2008c930e856045c8a/docs/easy_install.txt#L319

So I think this way of changing interpreter path in .py files should work in all cases and it definitely doesn't brake anything.
